aiida/tools/graph/graph_traversers.py

```
@@ -0,0 +1,287 @@
+        List with all the links that should be traversed in the backward direction.
+    """
+    # pylint: disable=too-many-locals,too-many-statements,too-many-branches
+    from aiida import orm
+    from aiida.tools.graph.age_entities import Basket
+    from aiida.tools.graph.age_rules import UpdateRule, RuleSequence, RuleSaveWalkers, RuleSetWalkers
+    from aiida.common import exceptions
+
+    if max_iterations is None:
+        max_iterations = inf
+    elif not (isinstance(max_iterations, int) or max_iterations is inf):
+        raise TypeError('Max_iterations has to be an integer or infinity')
+
+    linktype_list = []
+    for linktype in links_forward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_forward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_forwards = {'type': {'in': linktype_list}}
+
+    linktype_list = []
+    for linktype in links_backward:
+        if not isinstance(linktype, LinkType):
+            raise TypeError('links_backward should contain links, but one of them is: {}'.format(type(linktype)))
+        linktype_list.append(linktype.value)
+    filters_backwards = {'type': {'in': linktype_list}}
+
+    if not isinstance(starting_pks, (list, set, tuple)):
+        raise TypeError('starting_pks must be of type list, set or tuple\ninstead, it is {}'.format(type(starting_pks)))
+
+    if not starting_pks:
+        if get_links:
+            output = {'nodes': set(), 'links': set()}
+        else:
+            output = {'nodes': set(), 'links': None}
+        return output
+
+    operational_set = set()
+    for single_pk in starting_pks:
+        if not isinstance(single_pk, int):
+            raise TypeError(
+                'element provided in starting_pks cannot be pk because it is of type: {}'.format(type(single_pk))
+            )
+        operational_set.add(single_pk)
+
+    query_nodes = orm.QueryBuilder()
```

This might be a costly check. Do we really need it? Would ignoring non-existing pks be such a bad idea?

---

I'm usually wary of functions that simply decide that there is nothing to be done with your instructions without actually letting you know about this (basically helping to hide typos and errors in scripts). Maybe a compromise solution would be to add an input keyword 'ignore_absent' that is checked before doing the querying? (I would personally default it to `True` but I have less strong feelings about that)

---

Actually on second thought, the check is not that costly as it will only fetch the id's that are in the starting set, which typically is not that big. Secondly, I now actually agree with you that it is better not to silently swallow non-existing ids. So let's keep it like it was (you can remove the `ignore_absent` again, sorry .... :)

---

No probs, already done.